actors/actors.go

```diff
@@ -57,9 +91,89 @@ func (a *BuiltinActors) GetActorNameFromCid(actorCode cid.Cid) (string, error) {
 }
 
 func (a *BuiltinActors) GetActorCid(name string) (cid.Cid, error) {
-	if cid, ok := a.Metadata.ActorsNameCidMap[name]; ok {
+	if cid, ok := a.Metadata.ActorsNameCidMapByVersion[a.Metadata.Version][name]; ok {
+		return cid, nil
+	}
+
+	return cid.Cid{}, fmt.Errorf("actor '%s' not found in metadata", name)
+}
+
+func (a *BuiltinActors) GetActorNameFromCidByVersion(actorCode cid.Cid, version network.Version) (string, error) {
+	// Try the latest actors' version first
+	for name, code := range a.Metadata.ActorsNameCidMapByVersion[version] {
+		if actorCode == code {
+			return name, nil
+		}
+	}
+
+	// Try legacy actors
+	name := builtin.ActorNameByCode(actorCode)
+	if name != UnknownStr {
+		return name, nil
+	}
+
+	return UnknownStr, fmt.Errorf("invalid actor code CID: %s", actorCode)
+}
+
+func (a *BuiltinActors) GetActorCidByVersion(name string, version network.Version) (cid.Cid, error) {
+	if cid, ok := a.Metadata.ActorsNameCidMapByVersion[version][name]; ok {
 		return cid, nil
 	}
 
 	return cid.Cid{}, fmt.Errorf("actor '%s' not found in metadata", name)
 }
+
+func loadActorCids(currentNetworkVersion network.Version, loadAllActorVersions bool, lotusApi api.FullNode) (map[network.Version]ActorCidMap, error) {
+	var (
+		numWorkers      = 5
+		networkVersions []network.Version
+		actorCidsMap    = make(map[network.Version]ActorCidMap)
+	)
+
+	if loadAllActorVersions {
+		zap.S().Info("loading all actor versions")
+		networkVersions = make([]network.Version, LatestVersion+1)
+		for i := network.Version0; i <= LatestVersion; i++ {
+			networkVersions[i] = i
+		}
+	} else {
+		zap.S().Info("loading current actor version")
+		networkVersions = []network.Version{currentNetworkVersion}
+	}
+
+	versionChannel := make(chan network.Version, len(networkVersions))
+	actorCidsChannel := make(chan map[network.Version]ActorCidMap)
+
+	for i := 0; i < numWorkers; i++ {
+		go func(i int) {
+			for version := range versionChannel {
+				// todo: retry on network failure
+				actorCids, err := lotusApi.StateActorCodeCIDs(context.Background(), version)
+				if err != nil {
+					zap.S().Errorf("worker %d: error loading actor cids for version %d: %s", i, version, err.Error())
+					actorCids = ActorCidMap{}
```

Comment:
I would not do this. Otherwise, we end up hiding the error, and it will generate other errors in cascade. What if we just fail, return the error, and exit?